Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,498 @@
+            switch request.scheme {
+            case "http":
+                self.scheme = .http
+            case "https":
+                self.scheme = .https
+            default:
+                fatalError("HTTPClient.Request scheme should already be a valid one")
+            }
+            self.port = request.port
+            self.host = request.host
+        }
+
+        var scheme: Scheme
+        var host: String
+        var port: Int
+
+        enum Scheme: Hashable {
+            case http
+            case https
+        }
+    }
+
+    /// A `Connection` represents a `Channel` in the context of the connection pool
+    ///
+    /// In the `ConnectionPool`, each `Channel` belongs to a given `HTTP1ConnectionProvider`
+    /// and has a certain "lease state" (see the `isLeased` property).
+    /// The role of `Connection` is to model this by storing a `Channel` alongside its associated properties
+    /// so that they can be passed around together.
+    class Connection {
+        init(key: Key, channel: Channel, parentPool: ConnectionPool) {
+            self.key = key
+            self.channel = channel
+            self.parentPool = parentPool
+        }
+
+        /// Release this `Connection` to its associated `HTTP1ConnectionProvider` in the parent `ConnectionPool`
+        ///
+        /// This is exactly equivalent to calling `.release(theProvider)` on `ConnectionPool`
+        ///
+        /// - Warning: This only releases the connection and doesn't take care of cleaning handlers in th
+        ///  `Channel` pipeline.
+        func release() {
+            self.parentPool.release(self)
+        }
+
+        /// The connection pool this `Connection` belongs to.
+        ///
+        /// This enables calling methods like `release()` directly on a `Connection` instead of
+        /// calling `pool.release(connection)`. This gives a more object oriented feel to the API
+        /// and can avoid having to keep explicit references to the pool at call site.
+        let parentPool: ConnectionPool
+
+        /// The `Key` of the `HTTP1ConnectionProvider` this `Connection` belongs to
+        ///
+        /// This lets `ConnectionPool` know the relationship between `Connection`s and `HTTP1ConnectionProvider`s
+        fileprivate let key: Key
+
+        /// The `Channel` of this `Connection`
+        ///
+        /// - Warning: Requests that lease connections from the `ConnectionPool` are responsible
+        /// for removing the specific handlers they added to the `Channel` pipeline before releasing it to the pool.
+        let channel: Channel
+
+        /// Wether the connection is currently leased or not
+        var isLeased: NIOAtomic<Bool> = .makeAtomic(value: false)
+
+        /// Indicates that this connection is about to close
+        var isClosing: Bool = false
+
+        /// Convenience property indicating wether the underlying `Channel` is active or not
+        var isActive: Bool {
```

Comment:
I'm fairly sure that this property doesn't really help us. We're operating from outside the EventLoop which means that this property can change under our feet essentially at random (from true to false). I think we should remove it.

Reply:
Do you mean outside of an EL thread? I think I'm only using this from `releaseAction` which I think is only called from the event loop.

Reply:
@adtrevor in that case, can you put a `channelEL.assertInEventLoop()` into `var isActive`?

Reply:
Adding this doesn't crash, but now that I think about it it's probably just because there isn't sufficient test coverage: I think this is called from the taskEL and the only reason why this doesn't crash is that they are probably the same in the tests.

Also I think yesterday I missed that you was saying its racy just not at the "logical" level but mostly at the level of the memory, right? Because looking at the implementation it doesn't seem to be an atomic, and I thought it was.

Reply:
Sorry, looked again and I think it is actually atomic, so if there is no memory issue, I'm not sure the fact that this can change anytime would be an issue for our specific use case: this property is merely used as an additional check just before leasing a connection, to try to detect that we aren't leasing a closed connection.

If the connection was to close just between this call and the time when its actually leased, it's true that the request task would end up with a closed connection, but it seems to me that in such a case we would have already done everything in our power to ensure the connection will not close at the wrong time: alongside checking `isActive == true`, we are also ensuring that `isClosing == false` (it would be set to true if we determined the connection wasn't going to be a `keep-alive` one).

If the connection still closes, it means the server choose to close it just at "the wrong moment" between two requests, which I agree can happen, but I don't think this is linked to the synchronisation of `isClosing` anymore, wdyt?

Reply:
@adtrevor oh cool, just as an additional check is fine. Let's rename it then to something like

    var isActiveEstimation: Bool

that's more accurate.